gsgmail_connector.py

```diff
@@ -407,6 +338,148 @@ def _handle_run_query(self, param):
+        return {
+            "name": "Email Attachment Artifact",
+            "container_id": self.get_container_id(),
+            "cef": {
+                "vaultId": attach_resp[phantom.APP_JSON_HASH],
+                "fileHash": attach_resp[phantom.APP_JSON_HASH],
+                "file_hash": attach_resp[phantom.APP_JSON_HASH],
+                "fileName": file_name,
+            },
+            "run_automation": False,
+            "source_data_identifier": None,
+        }
+
+    def _parse_email_details(self, part, email_details):
+        headers = self._get_email_headers_from_part(part)
+        # split out important headers (for output table rendering)
+        if headers.get("to"):
+            email_details["to"] = headers["to"]
+
+        if headers.get("from"):
+            email_details["from"] = headers["from"]
+
+        if headers.get("subject"):
+            email_details["subject"] = headers["subject"]
+
+        part_type = part.get_content_type()
+        if part_type == "text/plain":
+            email_details["plain_bodies"].append(self._body_from_part(part))
+        elif part_type == "text/html":
+            email_details["html_bodies"].append(self._body_from_part(part))
+
+        email_details["email_headers"].append(headers)
+
+    def _get_payload_content(self, part):
+        if part.get_content_type().startswith("message/"):
+            return part.get_payload(0).as_string()
+        return part.get_payload(decode=True)
+
+    def _extract_attachment(self, part, action_result):
+        attach_resp = None
+        file_name = part.get_filename()
+        try:
+            # Create vault item with attachment payload
+            attach_resp = Vault.create_attachment(
+                self._get_payload_content(part),
+                container_id=self.get_container_id(),
+                file_name=file_name,
+            )
+        except Exception as e:
+            return action_result.set_status(
+                phantom.APP_ERROR,
+                f"Unable to add attachment: {file_name} Error: {self._get_error_message_from_exception(e)}",
+            )
+        if attach_resp.get("succeeded"):
+            # Create vault artifact
+            ret_val, msg, _ = self.save_artifact(
+                self._create_artifact(file_name, attach_resp)
+            )
+            if phantom.is_fail(ret_val):
+                return action_result.set_status(
+                    phantom.APP_ERROR,
+                    f"Could not save artifact to container: {msg}",
+                )
+        return phantom.APP_SUCCESS
+
+    @staticmethod
+    def _is_attachment(part):
+        return "attachment" in str(part.get("Content-Disposition"))
+
+    def _init_detail_fields(self, email_details):
+        email_details["plain_bodies"] = []
+        email_details["html_bodies"] = []
+        email_details["email_headers"] = []
+
+    def _parse_email_bodies(self, email_details):
```

Comment:
`format`/`join` would better describe what this does

Reply:
agreed